doc/man3/EVP_PKEY_size.pod

```diff
@@ -0,0 +1,39 @@
+=pod
+
+=head1 NAME
+
+EVP_PKEY_size, EVP_PKEY_bits, EVP_PKEY_security_bits
+- EVP_PKEY information functions
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ int EVP_PKEY_size(const EVP_PKEY *pkey);
+ int EVP_PKEY_bits(const EVP_PKEY *pkey);
+ int EVP_PKEY_security_bits(const EVP_PKEY *pkey);
+
+=head1 DESCRIPTION
+
+EVP_PKEY_size() returns the size of the key in I<pkey> in bytes.
+
+EVP_PKEY_bits() returns the size of the key in I<pkey> in bits.
```

Comment:
I would change the wording here to something similar to:

```
EVP_PKEY_bits() returns the cryptographic length in bits
of the cryptosystem to which the key in I<pkey> belongs.
Notice that the definition of cryptographic length is specific
to the key cryptosystem.
```
The difference is that a reader could be misled to think that this is some kind of measure of how big a buffer to contain some encoding of this key should be (see the recent #10542): it is not, and actually for some cryptosystems (e.g. DSA, ECDSA, DH, ECDH) this function does not give an information about the specific length of what constitutes the public key or the private key encapsulated by `pkey` as a particular instance, but rather a property of the domain parameters that define the cryptosystem `pkey` belongs to.
Using an analogy, if `DSA-2048` was a class and the argument `pkey` in a call to `EVP_PKEY_bits(pkey)` an instance of that class, the return value would be `2048`, and it would be a class attribute, not a instance attribute.

This view can actually be extended to RSA as well because even though there aren't really "domain parameters" associated with a key, and the _cryptographic length_ comes from the length of the instance attribute `n`, by construction we generally pick first an arbitrary bit length from a conventional list `{1024, 2048, 3072, 4096, ...}` and then we look for two primes that gives us a modulus of the desired length, so, maybe a bit forcefully, the key length could be seen as a "domain parameter" that we pick to define all the security properties of the cryptosystem to which our key belongs.

Anyway, this is the difference I try to highlight when I talk about _cryptographic length_: it should serve as a warning to the reader that it might differ from the conventional definition of the _concrete_ bit length of something stored or transmitted by a digital system, and that the nature of the return value of `EVP_PKEY_bits()` is somewhat more abstract.

## Some nasty details

I believe it is important to state that the definition of what constitutes the _cryptographic length_ of a key is part of the specific definition of the cryptosystem, as it is an arguably arbitrary choice of the designers.

### DSA keys (applies almost identically to DH as well)

- OpenSSL for DSA keys returns `BN_num_bits(pkey->pkey.dsa->p)`
- The DSA design defines the _length_ of the key as _L_, the bit length of the public parameter `p`; another relevant bit length is _N_ pertaining to the public parameter `q`.
- The value `x` that constitutes the private key is chosen randomly from `{1 ... (q - 1)}`, so its bit length is between 1 and _N_
- The value `y = g^x mod(p)` that constitutes the actual public key has a bit length between 1 and _L_
- In the cryptosystem as a whole we do some math modulo `q` and some math modulo `p`; the most computationally intense operations are done in the multiplicative group defined by `p`, so _L_ can be considered a metric for the computation level required for operations belonging to this cryptosystem; the definition of _L_ as the _key length_ also allowed direct comparison with the _key length_ of RSA (as the alternative and more well established asymmetric cryptosystem)

Given a chosen key length _L_, the _N_ length is basically picked through a table.
_L_ as a length is more closely related to the number of bits to transfer on the wire to distribute a public key (ignoring the fact that most likely you will need to transfer every time also the associated parameters, thus more than doubling the actual bits of information transferred), and only indirectly to the bits of security, which instead is more closely related to _N_ (intuitively, limiting us to a brute-force attack, there are at most `2^N` different private keys to explore to break the DSA cryptosystem with a given set of parameters `(p, q)`).

The choice of _L_ over _N_ as a measure of the _length_ of the key is thus a somewhat arbitrary design decision.
And it is also evident that this _cryptographic length_ comes from the parameters associated with the argument of the `EVP_PKEY_bits()` call, rather then being directly related to the size/length required to encode the particular key passed to the function.

### RSA keys

- OpenSSL for RSA keys returns `BN_num_bits(pkey->pkey.rsa->n)`
- The RSA design defines that the _length_ of a key is the bit length of the modulus `n` associated with the key pair, the secret key is the triplet `(p, q, d)` with `p * q = n`, and the public key is the pair `(e, n)`
- Complexity here is asymmetric, you optimize either for fast encryptions or decryptions, but roughly one could use this definition of length as a measure of complexity, area, or memory requirements for the cryptosystem operations
- The brute-force attack here does not lead far, as by construction `bitlen(p)` and `bitlen(q)` differ by a few bits, so blind brute-force requires more or less `bitlen(n)/2` steps; the best sieving attacks have well studied complexity though, so one can apply a formula to the bit length of `n` and directly derive the bits of strenght of a key (but generally one just uses a table because anyway we pick well determined lengths for `n` as the preliminary step of key generation).

It is again evident though, that the bit length of `n` is not the size of the private key (as `(p, q, d)` is already on average around `1.5 * n` and then usually we actually store more stuff to do math tricks for faster computations) but is close enough to the public key (`e` is usually small, so `(n, e)` is close in length.
Anyway it is definitely not the size in bits of any encoding of the specific key passed as an argument, but again the _length_ of the abstract class to which that key belongs.

### EC keys

- OpenSSL returns `EC_GROUP_order_bits(EC_KEY_get0_group(pkey->pkey.ec))`
- According to FIPS-186-3, for ECDSA, the order of the base point of the elliptic curve is `n`; and the bit length of `n` is considered to be the key size.
- The complications start when we note that in the cryptosystem we do some operations modulo `n`, the order of the EC group generated by the base point, and some operations modulo `q`, where `q` is the size of the field upon which the elliptic curve group is built (this means that the coordinates of any point on the curve are elements of the field and their maximum bit length is `bitlen(q)`). To complicate matters, usually the names assigned to standardized curves include a number, which corresponds to `q` rather than `n`. The bit sizes of `q` and `n` are bound by the Hasse theorem so that at most they differ by 1, but this can still lead to confusion as a curve named `sect163r1` has _key length_ 162 and this can appear as an off-by-one mistake.
- Another potential source of confusion is that in ECDH/ECDSA, `n` is directly related to the size of the **private key** rather than the public key (the opposite of what happens for DSA/DH/RSA where the `key length` is more closely related to the public key):
  + the secret key `d` is picked randomly in the range `{1, ... , n-1}`, so at most it has bit length `n`
  + the public key `Q = d * G` is a point on the elliptic curve with affine coordinates `(x, y)`, each in the range `{1, ... , q-1}`, thus at most twice bit length `q` (well point encoding actually wastes some bits with padding and optionally saves half the bits if using point compression)

### Curve25519 and Curve448

- Curve25519 and Curve448 are defined in an entirely different manner, and while their functioning is based on elliptic curve arithmetic
- While they have mathematical domain parameters that can be mapped to the traditional EC keys, they are defined as individual cryptosystems so that X25519 is one cryptosystem with exactly one set of parameters, Ed25519 is another distinct cryptosystem with its own set of domain parameters (equivalent to those of X25519, but not exactly the same), and the same goes for X448 and Ed448. This means that every Ed25519 key pair in existence has the same well defined set of parameters, there aren't "flavours" (vs RSA/DSA that can have 1024, 2048, etc. flavours, or DH keys that must be generated under the same agreed domain parameters, or EC keys over this or that curve parameters).
- As a consequence of the last point, the key lenght and the bits of security are defined directly by the _operational_ description of these cryptosystems as fixed constants (even though they can be derived from the mathematical description/proof of the systems) so we get:

https://github.com/openssl/openssl/blob/a0e594d6b5265efd412d248bd12e9d5305370286/crypto/ec/ecx_meth.c#L20-L21

https://github.com/openssl/openssl/blob/a0e594d6b5265efd412d248bd12e9d5305370286/crypto/ec/ecx_meth.c#L25-L27

or fixed constant definitions in IETF standards: https://tools.ietf.org/html/rfc8032#section-8.5

### Post-Quantum Crypto

We don't support any yet, and the only thing I dare say is that again each class of cryptosystems will have its own definition of what is considered the _cryptographic length_ of a key.

## Conclusion

Documenting `EVP_PKEY_bits()` is tricky!

Reply:
Will reword

Reply:
Done